src/runtime_src/core/pcie/linux/system_linux.cpp

```diff
@@ -98,26 +110,76 @@ static std::string machine_info()
   return model;
 }
 
-static std::vector<std::weak_ptr<xrt_core::device_linux>> mgmtpf_devices(16); // fix size
-static std::vector<std::weak_ptr<xrt_core::device_linux>> userpf_devices(16); // fix size
-static std::map<xrt_core::device::handle_type, std::weak_ptr<xrt_core::device_linux>> userpf_device_map;
-
 }
 
 namespace xrt_core {
 
+void
+system_linux::
+register_driver(std::shared_ptr<pcidrv::pci_driver> driver)
+{
+  driver_list.push_back(driver);
+  if (driver->is_user())
+    driver->scan_devices(user_ready_list, user_nonready_list);
+  else
+    driver->scan_devices(mgmt_ready_list, mgmt_nonready_list);
```

Comment:
Move the push_back to the end so as not to change driver_list state if scan_devices throws?
Also as you are correctly passing a copy of the shared pointer you should use driver_list.emplace_back(std::move(driver))  but of course not until after using the `driver` variable.

Reply:
Sure, will fix.